codegen/src/next/mod.rs

```diff
@@ -1,13 +1,145 @@
+            params: content.parse_terminated(Parameter::parse)?,
+        })
+    }
+}
+
+impl Endpoint {
+    fn from_item_fn(item: &mut ItemFn) -> Option<Self> {
+        let function = item.sig.ident.clone();
+
+        let params = if let Some(attribute) = item.attrs.get(0) {
+            attribute.tokens.clone()
+        } else {
+            return None;
+        };
+
+        item.attrs.clear();
+
+        let params: Params = match parse2(params) {
+            Ok(params) => params,
+            Err(err) => {
+                emit_error!(
+                    err.span(),
+                    err;
+                    hint = "The endpoint takes a comma-separated list of keys and values: `endpoint(method = get, route = \"/hello\")`"
+                );
+                return None;
+            }
+        };
+
+        // use paren span for missing argument errors
+        let paren = params.paren_token;
+
+        if params.params.is_empty() {
+            emit_error!(
+                paren.span,
+                "missing endpoint arguments";
+                hint = "The endpoint takes two arguments: `endpoint(method = get, route = \"/hello\")`"
+            );
+            return None;
+        }
+
+        let mut route = None;
+        let mut method = None;
+
+        for Parameter { key, value, .. } in params.params {
+            let key_ident = key.clone();
+            match key.to_string().as_str() {
+                "method" => {
+                    if let Expr::Path(path) = value {
+                        method = Some(path.path.segments[0].ident.clone());
+                    };
+                }
+                "route" => {
+                    if let Expr::Lit(literal) = value {
+                        if let Some(Lit::Str(literal)) = Some(literal.lit) {
+                            route = Some(literal);
+                        }
+                    }
+                }
+                _ => {
+                    emit_error!(
+                        key_ident,
+                        "invalid endpoint argument";
+                        hint = "Only `method` and `route` are valid endpoint arguments."
+                    );
+                    return None;
+                }
+            }
+        }
+
+        let route = if let Some(route) = route {
+            route
+        } else {
+            emit_error!(
+                paren.span,
+                "no route provided";
+                hint = "Add a route to your endpoint: `#[shuttle_codegen::endpoint(method = get, route = \"/hello\")]`"
+            );
+            return None;
+        };
+
+        let method = if let Some(method) = method {
+            method
+        } else {
+            emit_error!(
+                paren.span,
+                "no method provided";
+                hint = "Add a method to your endpoint: `#[shuttle_codegen::endpoint(method = get, route = \"/hello\")]`"
+            );
+            return None;
```

Comment:
I wonder if we can rather set a `has_error` variable to get all the errors to display in one go.

Reply:
Good idea! I put it below the "no parameters" error, since that already explains that both method and route are needed in its hint. I also figured we may return early if the params syntax is invalid, or if there is no endpoint attribute. I can move it up though if we want to group the syntax errors in as well.